Content.Shared/Chemistry/Solution.cs

```diff
@@ -0,0 +1,273 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using Robust.Shared.Interfaces.Serialization;
+using Robust.Shared.Serialization;
+using Robust.Shared.Utility;
+using Robust.Shared.ViewVariables;
+
+namespace Content.Shared.Chemistry
+{
+    /// <summary>
+    ///     A solution of reagents.
+    /// </summary>
+    public class Solution : IExposeData, IEnumerable<Solution.ReagentQuantity>
+    {
+        // Most objects on the station hold only 1 or 2 reagents
+        [ViewVariables]
+        private List<ReagentQuantity> _contents = new List<ReagentQuantity>(2);
+
+        /// <summary>
+        ///     The calculated total volume of all reagents in the solution (ex. Total volume of liquid in beaker).
+        /// </summary>
+        [ViewVariables]
+        public int TotalVolume { get; private set; }
+
+        /// <summary>
+        ///     Constructs an empty solution (ex. an empty beaker).
+        /// </summary>
+        public Solution() { }
+
+        /// <summary>
+        ///     Constructs a solution containing 100% of a reagent (ex. A beaker of pure water).
+        /// </summary>
+        /// <param name="reagentId">The prototype ID of the reagent to add.</param>
+        /// <param name="quantity">The quantity in milli-units.</param>
+        public Solution(string reagentId, int quantity)
+        {
+            AddReagent(reagentId, quantity);
+        }
+
+        /// <inheritdoc />
+        public void ExposeData(ObjectSerializer serializer)
+        {
+            serializer.DataField(ref _contents, "reagents", new List<ReagentQuantity>());
+
+            if (serializer.Reading)
+            {
+                TotalVolume = 0;
+                foreach (var reagent in _contents)
+                {
+                    TotalVolume += reagent.Quantity;
+                }
+            }
+        }
+
+        /// <summary>
+        ///     Adds a given quantity of a reagent directly into the solution. 
+        /// </summary>
+        /// <param name="reagentId">The prototype ID of the reagent to add.</param>
+        /// <param name="quantity">The quantity in milli-units.</param>
+        public void AddReagent(string reagentId, int quantity)
+        {
+            if(quantity <= 0)
```

Comment:
Wouldn't it make more sense to throw an out of range exception if the value is negative?

Reply:
dunno, the function as-is won't break if you give it a negative number. The exception would wreck the stack, and the only purpose would be to notify the client that they are trying to add a negative number. The code can handle a negative number, so is it really a program-stopping exception?